src/main/java/ch/njol/skript/util/BlockDataUtils.java

```diff
@@ -0,0 +1,330 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *
+ * Copyright 2011-2017 Peter Güttinger and contributors
+ */
+package ch.njol.skript.util;
+
+import java.io.NotSerializableException;
+import java.io.StreamCorruptedException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.lang.math.NumberUtils;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.data.BlockData;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+
+import ch.njol.skript.Skript;
+import ch.njol.util.StringUtils;
+import ch.njol.yggdrasil.Fields;
+import ch.njol.yggdrasil.YggdrasilSerializable.YggdrasilExtendedSerializable;
+
+/**
+ * Utility class for modifying {@link BlockData} of {@link Block Blocks}
+ */
+public class BlockDataUtils {
+	
+	private static final Map<String, StateType> stateTypes = new HashMap<>();
+	
+	static {
+		// Load all variations of block state types
+		if (Skript.isRunningMinecraft(1, 13)) {
+			for (Material material : Material.values()) {
+				// If this material is not a block, let's continue.
+				if (!material.isBlock())
+					continue;
+				
+				BlockData blockData = material.createBlockData();
+				String blockDataString = blockData.getAsString();
+				
+				// If this block data does not have states, let's continue.
+				if (!blockDataString.contains("["))
+					continue;
+				
+				String stateString = blockDataString.replaceAll(".+\\[|]", "");
+				String[] splits = stateString.split(",");
```

Comment:
Perhaps there would be a better way for this, since ``stateString`` is being iterated in quadratic time? Not entirely sure, but maybe there could be a way to loop through ``stateString`` only once for this.

Reply:
Im not sure what you mean.
The state string is created for each material.